### PR TITLE
113 captcha implementation

### PR DESCRIPTION
**Pull request by stadnitski**
Hi!
Here's a proof of concept of the captcha implementation. It works in the Stream mode. Please check the commit messages. 
Wondering to hear your feedback regarding the changes. 

Also please check the new setting `ForbidOnFailure` and the `handleBanResponseSoft()`. It is not related to captcha, but I hope it will be clear why this is needed. 

Please do not judge too strict. This is my first experience with Golang, traefik plugins, and so on :-D

### CODE REVIEW

**Comment by stadnitski** [comment]
@maxlerebourg @mathieuHa looking forward for your feedback. Please note that this is a proof of concept.

File: README.md [hunk 1]
```diff
@@ -67,7 +67,13 @@ Only one instance of the plugin is *possible*.
...
   - string
   - default: ""
-  - Crowdsec LAPI key for the bouncer : **must be unique by service**. 
+  - Crowdsec LAPI key for the bouncer : **must be unique by service**.
+- CrowdsecLapiMachineId
+  - string
+  - Used in NON-`alone` mode, Machine ID for Crowdsec LAPI (from `/etc/crowdsec/local_api_credentials.yaml`)
+- CrowdsecLapiPassword
+  - string
+  - Used in NON-`alone` mode, password for Crowdsec LAPI (from `/etc/crowdsec/local_api_credentials.yaml`)
```
Comment (maxlerebourg) [inline README.md:76]: good point 👍

File: bouncer.go [hunk 2]
```diff
@@ -266,6 +315,157 @@ type Login struct {
...
+
+func handleBanResponseSoft(bouncer *Bouncer, rw http.ResponseWriter, req *http.Request) {
+	if bouncer.forbidOnFailure {
+		rw.WriteHeader(http.StatusForbidden)
+		return
+	}
+	bouncer.next.ServeHTTP(rw, req)
+}
+
+func handleBanResponseForced(bouncer *Bouncer, rw http.ResponseWriter, req *http.Request) {
```
Comment (maxlerebourg) [inline bouncer.go:339]: not understand why you create a function for one line

File: bouncer.go [hunk 3]
```diff
@@ -391,34 +616,63 @@ func handleStreamCache(bouncer *Bouncer) error {
...
 	isCrowdsecStreamHealthy = true
 	return nil
 }
 
-func crowdsecQuery(bouncer *Bouncer, stringURL string, isPost bool) ([]byte, error) {
+func crowdsecQuery(bouncer *Bouncer, stringURL string, method string) ([]byte, error) {
 	var req *http.Request
-	if isPost {
+	var machineId string
+	var password string
```
Comment (maxlerebourg) [inline bouncer.go:644]: you can declare these 2 variables inside the `if method == http.MethodPost`

File: pkg/captcha/captcha.html [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        </script>
+    </head>
+    <body>
+        <h1>Please verify you're not a robot</h1>
+        <div id="recaptcha"></div>
+        <script src="https://www.google.com/recaptcha/api.js?onload=onloadCallback&render=explicit"
+                async defer>
+        </script>
+    </body>
+</html>
```
Comment (maxlerebourg) [inline pkg/captcha/captcha.html:27]: EOF: you need to have a break line at the end of your file

File: bouncer.go [hunk 5]
```diff
@@ -221,16 +255,24 @@ func (bouncer *Bouncer) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	// Right here if we cannot join the stream we forbid the request to go on.
 	if bouncer.crowdsecMode == configuration.StreamMode || bouncer.crowdsecMode == configuration.AloneMode {
 		if isCrowdsecStreamHealthy {
+			logger.Debug(fmt.Sprintf("ServeHTTP:stream is healthy"))
 			bouncer.next.ServeHTTP(rw, req)
 		} else {
 			logger.Debug(fmt.Sprintf("ServeHTTP isCrowdsecStreamHealthy:false ip:%s", remoteIP))
-			rw.WriteHeader(http.StatusForbidden)
+			if bouncer.forbidOnFailure {
```
Comment (maxlerebourg) [inline bouncer.go:262]: this is your function `handleBanResponseSoft`

File: bouncer.go [hunk 6]
```diff
@@ -266,6 +315,157 @@ type Login struct {
...
+		logger.Debug(fmt.Sprintf("Error reading captcha HTML file '%s'", bouncer.captchaHtmlFilePath))
+	}
+
+	rw.Header().Add("Content-Type", "text/html")
+	rw.WriteHeader(200)
+	_, err = rw.Write(content)
+}
+
+func handleCaptchaValidation(bouncer *Bouncer, rw http.ResponseWriter, req *http.Request, remoteIP string) bool {
+	if !(req.RequestURI == captchaVerifyRoute && req.Method == "POST") {
```
Comment (maxlerebourg) [inline bouncer.go:366]: !(a && b) === !a || !b

File: bouncer.go [hunk 7]
```diff
@@ -266,6 +315,157 @@ type Login struct {
...
+func handleBanResponseOrCaptcha(bouncer *Bouncer, rw http.ResponseWriter, req *http.Request, isCaptcha bool) {
+	logger.Debug(fmt.Sprintf("handleBanResponseOrCaptcha '%t'", isCaptcha))
+	handleBanResponseForced(bouncer, rw, req)
+
+	if !isCaptcha {
+		return
+	}
+
+	logger.Debug(fmt.Sprintf("Must respond with captcha here.... Captcha file '%s'", bouncer.captchaHtmlFilePath))
+	content, err := os.ReadFile(bouncer.captchaHtmlFilePath)
```
Comment (maxlerebourg) [inline bouncer.go:353]: you can declare the captchaPage in the global context, not need to read the file every time you send it

File: bouncer.go [hunk 8]
```diff
@@ -22,20 +23,29 @@ import (
...
 
 //nolint:gochecknoglobals
 var (
 	isStartup               = true
 	isCrowdsecStreamHealthy = true
 	ticker                  chan bool
+	captchaSecretKey        = ""
+	captchaSiteKey          = ""
+	captchaVerifyRoute      = ""
+	jwtLapiToken            = ""
```
Comment (maxlerebourg) [inline bouncer.go:48]: You declare variables globaly, it's fine but you use bouncer's one.
Choose one way of doing this.
Prefer the bouncer one `//nolint:gochecknoglobals` GO don't really like global vars

File: bouncer.go [hunk 9]
```diff
@@ -391,34 +616,63 @@ func handleStreamCache(bouncer *Bouncer) error {
...
 		))
 		req, _ = http.NewRequest(http.MethodPost, stringURL, bytes.NewBuffer(data))
 	} else {
-		req, _ = http.NewRequest(http.MethodGet, stringURL, nil)
+		req, _ = http.NewRequest(method, stringURL, nil)
 	}
+
+	if method == http.MethodDelete {
+		bearer := fmt.Sprintf("Bearer %s", jwtLapiToken)
+		logger.Debug(fmt.Sprintf("BEARER: '%s'", bearer))
```
Comment (maxlerebourg) [inline bouncer.go:668]: Hummm could be dangerous to log the token

File: bouncer.go [hunk 10]
```diff
@@ -362,27 +562,52 @@ func getToken(bouncer *Bouncer) error {
 	return fmt.Errorf("getToken statusCode:%d", login.Code)
 }
 
+func refreshLapiToken(bouncer *Bouncer) {
```
Comment (maxlerebourg) [inline bouncer.go:565]: Customize the getToken function to return a string that can be the CAPI_CrowdsecKey or LAPI_JWT, and in params the url
func get(bouncer, url) (string, error) {...}
That will save us from copy paste function refreshLapiToken

**Comment by maxlerebourg** [comment]
I spent a long time on the review, don't take everything personally, I tried to explain as much as possible my point of view.

Thank you a LOT for your interest in this project.
I'll try to test your implementation as soon as possible.

**Comment by stadnitski** [comment]
> I spent a long time on the review, don't take everything personally, I tried to explain as much as possible my point of view.
> 
> Thank you a LOT for your interest in this project. I'll try to test your implementation as soon as possible.

Hey, thanks for the feedback, it make sense, nothing is taken personally. Here's my bouncer config for the nginx service, which can help with testing: 

```
        # Definition of the middleware
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.enabled=true
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecMode=stream
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.updateIntervalSeconds=5
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.logLevel=DEBUG
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiKey=582a29e4a15d13cf74ad382db4a7de2e
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiScheme=http
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiHost=crowdsec-lapi:8080
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiMachineID=crowdsec-lapi
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiPassword=<THE PASSWORD>
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.forbidOnFailure=false
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.captchaHtmlFilePath=/captcha.html
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.captchaSiteKey=<SITE KEY HERE>
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.captchaSecretKey=<SECRET KEY HERE>
        - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.captchaVerifyRoute=/not-a-robot

```

**Comment by mathieuHa** [comment]
> > I spent a long time on the review, don't take everything personally, I tried to explain as much as possible my point of view.
> > Thank you a LOT for your interest in this project. I'll try to test your implementation as soon as possible.
> 
> Hey, thanks for the feedback, it make sense, nothing is taken personally. Here's my bouncer config for the nginx service, which can help with testing:
> 
> ```
>         # Definition of the middleware
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.enabled=true
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecMode=stream
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.updateIntervalSeconds=5
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.logLevel=DEBUG
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiKey=582a29e4a15d13cf74ad382db4a7de2e
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiScheme=http
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiHost=crowdsec-lapi:8080
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiMachineID=crowdsec-lapi
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.crowdsecLapiPassword=<THE PASSWORD>
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.forbidOnFailure=false
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.captchaHtmlFilePath=/captcha.html
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.captchaSiteKey=<SITE KEY HERE>
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.captchaSecretKey=<SECRET KEY HERE>
>         - traefik.http.middlewares.crowdsec-bouncer.plugin.bouncer.captchaVerifyRoute=/not-a-robot
> ```

Good Point, we will definitely need to make a complete example of this feature, in the folder examples: maybe captcha-ban

File: README.md [hunk 11]
```diff
@@ -133,6 +139,23 @@ Only one instance of the plugin is *possible*.
...
+  - Used to specify the location of the captcha HTML file to be sent in the response if Crowdsec decision is to show captcha.
+- captchaSiteKey
+  - string
+  - Google reCaptcha Site Key
+- captchaSecretKey
+  - string
+  - Google reCaptcha Secret Key
+- captchaVerifyRoute
+  - string
+  - dafault: `not-a-robot`
```
[the remark this answers is not in the document]
Comment (mathieuHa) [reply README.md:157]: Maj stands for capital letter

**Comment by maxlerebourg** [comment]
Hey @stadnitski, have you plan to fix all the point I raised ?

**Comment by stadnitski** [comment]
Hey @maxlerebourg, yes, it's done already, but still got to find a little time to finish a few tests and push the changes. Let's see if that can be done today.

File: bouncer.go (hunk 2 above)
Comment (stadnitski) [reply bouncer.go:339]: this behaviour is required multiple times, which means it got to be a function

File: bouncer.go (hunk 5 above)
Comment (stadnitski) [reply bouncer.go:262]: done, coming soon

File: bouncer.go (hunk 6 above)
Comment (stadnitski) [reply bouncer.go:366]: won't !a || !b return true if either of them is true? 
!(a && b) will return true when both a and b are false. 
Here we want to exit from the function if the URI is not out "verification URI", AND if method is not POST.

Comment (stadnitski) [reply bouncer.go:366]: won't !a || !b return true if either of them is true? 
!(a && b) will return true when both a and b are false. 
Here we want to exit from the function if the URI is not out "verification URI", AND if method is not POST.

File: bouncer.go (hunk 8 above)
Comment (stadnitski) [reply bouncer.go:48]: good point, thanks!

File: bouncer.go (hunk 10 above)
Comment (stadnitski) [reply bouncer.go:565]: Yep, good point. Done.

File: bouncer.go (hunk 6 above)
Comment (maxlerebourg) [reply bouncer.go]: so use !a && !b

File: pkg/cache/cache.go [hunk 12]
```diff
@@ -27,13 +26,16 @@ var (
...
 
-func (localCache) getDecision(clientIP string) (bool, error) {
+func (localCache) getDecision(clientIP string) (bool, bool, error) {
 	banned, isCached := cache.Get(clientIP)
 	bannedString, isValid := banned.(string)
+	logger.Debug(fmt.Sprintf("cache:getDecision:bannedString '%s', isValid '%t', isCached '%t'", bannedString, isValid, isCached))
 	if isCached && isValid && len(bannedString) > 0 {
-		return bannedString == cacheBannedValue, nil
+		isBanned := bannedString == cacheBannedValue || bannedString == cacheCaptchaValue
+		isCaptcha := bannedString == cacheCaptchaValue
```
Comment (maxlerebourg) [inline pkg/cache/cache.go:35]: isBanned := bannedString != cacheNoBannedValue
isCaptcha := bannedString == cacheCaptchaValue

or  

isCaptcha := bannedString == cacheCaptchaValue
isBanned := bannedString == cacheBannedValue || isCaptcha

Comment (stadnitski) [reply pkg/cache/cache.go:35]: Oh, didn't know Go can do that, thanks :)

**Comment by stadnitski** [comment]
Hey @maxlerebourg @mathieuHa 
Need your opinion and help with the idea I got in the last commit. 

Let's imagine a situation when infra has got 3 Traefik nodes, the bouncer is in either:
* live mode with 2 minute cache 
* stream mode with 2 minute update
So the user comes to the website, where his IP is banned with captcha decision. 
The used proves he's not a robot, and this will trigger the ban decision removal from LAPI. 
The Traefik nodes might have to get know about the changes in a while, and we must let the user in immediately. So the idea is to use a cookie, which will be an encoded "something" that we could decode and perform verification. 

Suggested are two functions `createBypassCookie` and `validateByPassCookie`.  
Any ideas about how they could be implemented? Or maybe someone would like to take that over?

**Comment by mathieuHa** [comment]
> Hey @maxlerebourg @mathieuHa Need your opinion and help with the idea I got in the last commit.
> 
> Let's imagine a situation when infra has got 3 Traefik nodes, the bouncer is in either:
> 
> * live mode with 2 minute cache
> * stream mode with 2 minute update
>   So the user comes to the website, where his IP is banned with captcha decision.
>   The used proves he's not a robot, and this will trigger the ban decision removal from LAPI.
>   The Traefik nodes might have to get know about the changes in a while, and we must let the user in immediately. So the idea is to use a cookie, which will be an encoded "something" that we could decode and perform verification.
> 
> Suggested are two functions `createBypassCookie` and `validateByPassCookie`. Any ideas about how they could be implemented? Or maybe someone would like to take that over?

Hi indeed, that behing banned while we juste solved a captcha would be a weird user experience.  

Adding a cookie to the user browser that will **bypass crowdsec** check would mean **anyone can bypass bans and crowdsec checks** as cookies can be forged user side to any value.  


I'd like to review couples possible cases that could happen:

##### Case 1 - 2 nodes **A**(`stream mode`, **B** `stream mode`)

First Query goes to node A, user solve captcha, plugin sends information to the LAPI, but it also update his cache (local or redis) at the same time
Second Query goes to node A, user already unbanned from cache no issue
Third  Query goes to node B, local cache would be an issue if stream has not updated but redis would not if shared redis


##### Case 2 -  2 nodes A(`live mode`, B `live mode`)

First Query goes to node A, user solve captcha, plugin sends information to the LAPI
Second Query goes to node A, which asks LAPI and goes through
Third  Query goes to node B, which asks LAPI and goes through

_No issue here_

##### Case 3 - 2 nodes **A**(`stream mode`, **B** `live mode`)

First Query goes to node A, user solve captcha, plugin sends information to the LAPI, but it also update his cache (local or redis) at the same time
Second Query goes to node A, user already unbanned from cache no issue
Third  Query goes to node B, which asks LAPI and goes through

_No issue here_

##### Case 4 - 2 nodes **B**(`live mode`, **B** `stream mode`)

First Query goes to node A, user solve captcha, plugin sends information to the LAPI
Second Query goes to node A, which asks LAPI and goes through
Third  Query goes to node B, local cache would be an issue if stream has not updated

**To sum up**, the only issue I see is when a user goes through one node and then another, if the local cache has not sync with LAPI in between it will ban the user again.

In the worst case, the user would have to solve another captcha for each new node it manages to go through in between sync.

I think it can be totally avoided by always routing the user through the same node, which I think can be done easily with [sticky sessions](https://traefik.io/glossary/what-are-sticky-sessions/).

However, it is imported to inform the Traefik admin that is needs to be enabled in order to avoid this issue.

@maxlerebourg @stadnitski Let me know what you think

**Comment by maxlerebourg** [comment]
For me, the only solution for this configuration (multiple Traefiks in any mode) is to use a shared Redis cache.
Local storage is not recommended when there is more than one Traefik.

If you find the cookie solution implemented in the official NGINX bouncer, we could think about it, otherwise we can't approve this behavior, sorry @stadnitski.
We need to follow the KISS principle (Keep it simple, stupid), and the cookie solution seems too complex, and with a lot of vulnerabilities like @mathieuHa said.

File: pkg/cache/cache.go [hunk 13]
```diff
@@ -27,13 +26,16 @@ var (
 
 type localCache struct{}
 
-func (localCache) getDecision(clientIP string) (bool, error) {
+func (localCache) getDecision(clientIP string) (bool, bool, error) {
```
Comment (maxlerebourg) [inline pkg/cache/cache.go:29]: what do you think to expose these 3 variables (Capitalize the first letter) and this function return the bannedString like `func (localCache) getDecision(clientIP string) (string, error) {`
CacheBannedValue   = "t"
CacheNoBannedValue = "f"
CacheCaptchaValue  = "c"

**Comment by stadnitski** [comment]
@maxlerebourg @mathieuHa hey mates. Sorry for being away for such a long time. Had a little work-related travel. Now I'm back and got to finish the bouncer implementation into our infra. I can see there's another implementation coming from @ LaurenceJJones. Should we somehow decide which way we're rawing together?

File: pkg/cache/cache.go (hunk 13 above)
Comment (stadnitski) [reply pkg/cache/cache.go:29]: that make sense

**Comment by mathieuHa** [comment]
> @maxlerebourg @mathieuHa hey mates. Sorry for being away for such a long time. Had a little work-related travel. Now I'm back and got to finish the bouncer implementation into our infra. I can see there's another implementation coming from @ LaurenceJJones. Should we somehow decide which way we're rawing together?

Hey @stadnitski @LaurenceJJones,

Sorry for the delay, I know this has been opened for some time now with 2 PR.
We will try with @maxlerebourg to take some time review them and see if we can progress on this based on both ideas/proposition.

**Comment by stadnitski** [comment]
> > @maxlerebourg @mathieuHa hey mates. Sorry for being away for such a long time. Had a little work-related travel. Now I'm back and got to finish the bouncer implementation into our infra. I can see there's another implementation coming from @ LaurenceJJones. Should we somehow decide which way we're rawing together?
> 
> Hey @stadnitski @LaurenceJJones,
> 
> Sorry for the delay, I know this has been opened for some time now with 2 PR. We will try with @maxlerebourg to take some time review them and see if we can progress on this based on both ideas/proposition.

Hey @mathieuHa, looking forward for hearing back from you :-)

**Comment by mathieuHa** [comment]
> > > @maxlerebourg @mathieuHa hey mates. Sorry for being away for such a long time. Had a little work-related travel. Now I'm back and got to finish the bouncer implementation into our infra. I can see there's another implementation coming from @ LaurenceJJones. Should we somehow decide which way we're rawing together?
> > 
> > 
> > Hey @stadnitski @LaurenceJJones,
> > Sorry for the delay, I know this has been opened for some time now with 2 PR. We will try with @maxlerebourg to take some time review them and see if we can progress on this based on both ideas/proposition.
> 
> Hey @mathieuHa, looking forward for hearing back from you :-)

@maxlerebourg let's review it this weekend

**Comment by mathieuHa** [comment]
Hi @stadnitski @LaurenceJJones,

We took some time this weekend with @maxlerebourg to review the two proposal and used thosed ideas to do an implementation of the catpcha remediation process.
I just released a [beta version](https://github.com/maxlerebourg/crowdsec-bouncer-traefik-plugin/releases/tag/v1.3.0-beta1)
We will continue to test it and expect some feedback from users to see if it works as it should.
Thanks for your time spent on this!

Regards,

**Comment by stadnitski** [comment]
Thanks @mathieuHa, @maxlerebourg and @LaurenceJJones
I will be taking this to production in late April or May. Feedback will follow.
